### ISSUE TITLE
Deserialize snapshot and inheritance

### ISSUE THREAD

**Issue by Codeluck** [charto/classy-mst#3]
Hi there!
I got an issue I couldn't solve. Please, give me some suggestion. One of key thing of classy-mst is the support of inheritance. However, it doesn't work after saving an MST node as a snapshot and then deserialize it back. For. ex.: I have base Layer class and another one is TextLayer inherited from Layer. I wanna serialize TextLayer and then at certain moment I would like to deserialize it back. But the system has no idea about its type. How to avoid this issue? Do I need to save extra type information somewhere?

**Comment by jjrv** [comment]
You're right that extra type information is needed. `classy-mst` should handle it using it dispatcher and `postProcessSnapshot` hook but currently doesn't. I need to fix this.

**Comment by Codeluck** [comment]
Any progress on that? I really need this feature in my project. What's your approach to serialize extra type information?

**Comment by jjrv** [comment]
If you have a list of layers like `types.array(Layer)` then it won't accept a `TextLayer`. Instead it needs to be `types.array(types.union(dispatcher, Layer, TextLayer))`. For polymorphism to work as expected, every time a parent class is mentioned somewhere, it needs to be replaced with a union including all of its subclasses. It seems `classy-mst` could do this automatically using late-bound types for all classes. The downside is that all types must be declared before any type instances are created.

Still working on this, lots of other urgent things going on unfortunately.

**Comment by jjrv** [comment]
Can you test with the current version in the `develop` branch? Just remove `classy-mst` from `node_modules` and then inside issue the commands:

```bash
git clone https://github.com/charto/classy-mst.git
cd classy-mst
git checkout develop
npm install
```

The `mst` function now takes a third, optional parameter: the class name. It gets stored in the snapshot under a key named `$`. If you wish to rename the key, then before creating any model instances use:

```TypeScript
import { setTypeTag } from 'classy-mst';
setTypeTag('my-key-name');
```

**Comment by jjrv** [comment]
This should be fixed in version 0.3.0 released just now.
Feel free to reopen if there are still problems.
